File: src/test/java/net/flyingfishflash/loremlist/integration/domain/item/ItemUpdateTest.java
package net.flyingfishflash.loremlist.integration.domain.item;

import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.stream.Stream;
import net.flyingfishflash.loremlist.api.data.request.LrmItemCreateRequest;
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfProblem;
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfSuccess;
import net.flyingfishflash.loremlist.domain.lrmitem.ItemNotFoundException;
import net.flyingfishflash.loremlist.integration.domain.AbstractIntegrationTest;
import net.flyingfishflash.loremlist.integration.domain.TestData;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ItemUpdateTest extends AbstractIntegrationTest {

  private record ListItemRequestRecord(
      UUID listId, UUID itemId, LrmItemCreateRequest listItemCreateRequest) {}

  private List<ListItemRequestRecord> requestsAlpha;

  @BeforeAll
  void setUp() throws Exception {
    purgeDomain();
    UUID listWithItems = createAndVerifyList(TestData.LIST_CREATE_REQUESTS.get(0));

    requestsAlpha = new ArrayList<>();
    for (var request : TestData.ITEM_CREATE_REQUESTS_ALPHA) {
      requestsAlpha.add(
          new ListItemRequestRecord(
              listWithItems, createAndVerifyListItem(listWithItems, request), request));
    }

    verifyContentValue("/lists/count", 1);
    verifyContentValue("/lists/" + listWithItems + "/items/count", requestsAlpha.size());
  }

  @Test
  void failsWhenItemToUpdateIsNotFound() throws Exception {
    UUID nonExistentUuid = UUID.randomUUID();
    verifyRequest(HttpMethod.PATCH, "/items/" + nonExistentUuid)
        .requestBody("{ \"name\": null, \"description\": null }")
        .expectedDisposition(DispositionOfProblem.FAILURE)
        .statusMatcher(status().isNotFound())
        .additionalMatchers(
            jsonPath("$.content.title").value(ItemNotFoundException.class.getSimpleName()),
            jsonPath("$.content.status").value("404"))
        .perform();
  }

  @Test
  void failsWhenRequestBodyIncludesUnsupportedField() throws Exception {
    verifyRequest(HttpMethod.PATCH, "/items/" + requestsAlpha.get(0).itemId())
        .requestBody(
            "{ \"name\": \"lorem ipsum\", \"description\": \"lorem ipsum\", \"zzz\": \"blah\" }")
        .expectedDisposition(DispositionOfProblem.ERROR)
        .statusMatcher(status().isInternalServerError())
        .additionalMatchers(
            jsonPath("$.content.title").value(IllegalArgumentException.class.getSimpleName()),
            jsonPath("$.content.status").value(HttpStatus.INTERNAL_SERVER_ERROR.value()))
        .perform();
  }

  @ParameterizedTest(name = "succeeds for all supported fields of ''{0}''")
  @MethodSource("requestsAlpha")
  void succeedsForAllSupportedFields(String name, ListItemRequestRecord record) throws Exception {
    String newName = record.listItemCreateRequest().getName() + " *";
    String newDescription = record.listItemCreateRequest().getDescription() + " *";
    verifyRequest(HttpMethod.PATCH, "/items/" + record.itemId())
        .requestBody(
            "{ \"name\": \"" + newName + "\", \"description\": \"" + newDescription + "\" }")
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .additionalMatchers(
            jsonPath("$.content.id").value(record.itemId().toString()),
            jsonPath("$.content.name").value(newName),
            jsonPath("$.content.description").value(newDescription),
            jsonPath("$.content.created").isNotEmpty(),
            jsonPath("$.content.updated").isNotEmpty())
        .perform();

    verifyRequest(HttpMethod.GET, "/items/" + record.itemId())
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .additionalMatchers(
            jsonPath("$.content.name").value(newName),
            jsonPath("$.content.description").value(newDescription))
        .perform();
  }

  Stream<Arguments> requestsAlpha() {
    return requestsAlpha.stream().map(r -> Arguments.of(r.listItemCreateRequest().getName(), r));
  }
}

File: src/test/kotlin/net/flyingfishflash/loremlist/integration/domain/item/ItemUpdateTest.kt
package net.flyingfishflash.loremlist.integration.domain.item

import io.kotest.datatest.withData
import net.flyingfishflash.loremlist.api.data.request.LrmItemCreateRequest
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfProblem
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfSuccess
import net.flyingfishflash.loremlist.domain.lrmitem.ItemNotFoundException
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest.TestData.itemCreateRequestsAlpha
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest.TestData.listCreateRequests
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Transactional
class ItemUpdateTest(mockMvc: MockMvc) :
  DomainFunctionTest(mockMvc = mockMvc, body = {

    data class ListItemRequestRecord(val listId: UUID, val itemId: UUID?, val listItemCreateRequest: LrmItemCreateRequest)
    var requestsAlpha: List<ListItemRequestRecord> = emptyList()
    beforeSpec {
      purgeDomain()

      // create a list
      val listWithItems = createAndVerifyList(listRequest = listCreateRequests[0])

      // create list items
      requestsAlpha = itemCreateRequestsAlpha.map {
        ListItemRequestRecord(listId = listWithItems, itemId = createAndVerifyListItem(listWithItems, it), listItemCreateRequest = it)
      }

      // validate state
      verifyContentValue(url = "/lists/count", expectedValue = 1)
      verifyContentValue(url = "/lists/$listWithItems/items/count", expectedValue = requestsAlpha.size)
    }

    describe("item modification") {
      context("modifying an item: /items/{item-id}") {
        it("fails when item to update is not found") {
          val nonExistentUuid = UUID.randomUUID()

          performRequestAndVerifyResponse(
            method = HttpMethod.PATCH,
            instance = "/items/$nonExistentUuid",
            requestBody = "{ \"name\": null, \"description\": null }",
            expectedDisposition = DispositionOfProblem.FAILURE,
            statusMatcher = status().isNotFound(),
            additionalMatchers = arrayOf(
              jsonPath("$.content.title").value(ItemNotFoundException::class.java.simpleName),
              jsonPath("$.content.status").value("404"),
            ),
          )
        }

        it("fails when request body includes unsupported field") {
          performRequestAndVerifyResponse(
            method = HttpMethod.PATCH,
            instance = "/items/${requestsAlpha.first().itemId}",
            requestBody = "{ \"name\": \"lorem ipsum\", \"description\": \"lorem ipsum\", \"zzz\": \"blah\" }",
            expectedDisposition = DispositionOfProblem.ERROR,
            statusMatcher = status().isInternalServerError,
            additionalMatchers = arrayOf(
              jsonPath("$.content.title").value(IllegalArgumentException::class.java.simpleName),
              jsonPath("$.content.status").value(HttpStatus.INTERNAL_SERVER_ERROR.value()),
            ),
          )
        }

        withData(
          nameFn = { record -> "succeeds for all supported fields of '${record.listItemCreateRequest.name}'" },
          requestsAlpha,
        ) { record ->
          performRequestAndVerifyResponse(
            method = HttpMethod.PATCH,
            instance = "/items/${record.itemId}",
            requestBody = "{ \"name\": \"${record.listItemCreateRequest.name} *\", \"description\": \"${record.listItemCreateRequest.description} *\" }",
            expectedDisposition = DispositionOfSuccess.SUCCESS,
            additionalMatchers = arrayOf(
              jsonPath("$.content.id").value(record.itemId.toString()),
              jsonPath("$.content.name").value("${record.listItemCreateRequest.name} *"),
              jsonPath("$.content.description").value("${record.listItemCreateRequest.description} *"),
              jsonPath("$.content.created").isNotEmpty(),
              jsonPath("$.content.updated").isNotEmpty(),
            ),
          )

          performRequestAndVerifyResponse(
            method = HttpMethod.GET,
            instance = "/items/${record.itemId}",
            expectedDisposition = DispositionOfSuccess.SUCCESS,
            additionalMatchers = arrayOf(
              jsonPath("$.content.name").value("${record.listItemCreateRequest.name} *"),
              jsonPath("$.content.description").value("${record.listItemCreateRequest.description} *"),
            ),
          )
        }
      }
    }
  })
